examples/Nat/TcpState.cs

```diff
@@ -0,0 +1,126 @@
+/*
+Pax : tool support for prototyping packet processors
+Jonny Shipton, Cambridge University Computer Lab, July 2016
+
+Use of this source code is governed by the Apache 2.0 license; see LICENSE.
+*/
+
+using System;
+using PacketDotNet;
+
+namespace Pax.Examples.Nat
+{
+  /// <summary>
+  /// Represents the state of a TCP connection.
+  /// </summary>
+  internal class TcpState : ITransportState<TcpPacket>
+  {
+    /// <summary>
+    /// The time for which a TCP connection should be kept open after the remote host terminates it with a Fin.
+    /// </summary>
+    private readonly TimeSpan TIME_WAIT;
+
+    /// <summary>
+    /// The inferred state of the TCP connection from the inside node to the outside node. Related to the TCP state on the outside node.
+    /// </summary>
+    private TcpDirectionalState InOutConnection = TcpDirectionalState.None;
+
+    /// <summary>
+    /// The inferred state of the TCP connection from the outside node to the inside node. Related to the TCP state on the inside node.
+    /// </summary>
+    private TcpDirectionalState OutInConnection = TcpDirectionalState.None;
+
+    /// <summary>
+    /// The time that the connection was closed, if applicable.
+    /// </summary>
+    private DateTime? CloseTime = null;
+
+    /// <summary>
+    /// True if the TCP connection from the inside to the outside is closed.
+    /// </summary>
+    public bool ClosedFromInside { get { return InOutConnection == TcpDirectionalState.None || InOutConnection == TcpDirectionalState.FinAck; } }
+
+    /// <summary>
+    /// True if the TCP connection from the outside to the inside is closed.
+    /// </summary>
+    public bool ClosedFromOutside { get { return OutInConnection == TcpDirectionalState.None || OutInConnection == TcpDirectionalState.FinAck; } }
+
+    /// <summary>
+    /// True if the connection is in the TIME_WAIT state.
+    /// </summary>
+    public bool InTimeWait { get { return CloseTime.HasValue && DateTime.Now - CloseTime < TIME_WAIT; } }
+
+    /// <summary>
+    /// True if the TCP connections in both directions are closed, determining if the connection can be removed before the timeout elapses.
+    /// </summary>
+    public bool CanBeClosed { get { return ClosedFromInside && ClosedFromOutside && !InTimeWait; } }
+
+    /// <summary>
+    /// Creates a new TcpState in the initial state.
+    /// </summary>
+    /// <param name="time_wait">The duration of the TIME_WAIT state.</param>
+    public TcpState(TimeSpan time_wait)
+    {
+      TIME_WAIT = time_wait;
+    }
+
+    public void UpdateState(TcpPacket packet, bool packetFromInside)
```

Comment:
Thanks! It would be very useful to have this documented as a comment, 
since it exposes the limitations of the system and flags up possible 
improvements.

On 2016-07-18 15:11, TMVector wrote:
> In examples/Nat/TcpState.cs [1]:
> 
>> +
>> + /// <summary>
>> + /// True if the TCP connections in both directions are closed,
> determining if the connection can be removed before the timeout
> elapses.
>> + /// </summary>
>> + public bool CanBeClosed { get { return ClosedFromInside &&
> ClosedFromOutside && !InTimeWait; } }
>> +
>> + /// <summary>
>> + /// Creates a new TcpState in the initial state.
>> + /// </summary>
>> + /// <param name="time_wait">The duration of the TIME_WAIT
> state.</param>
>> + public TcpState(TimeSpan time_wait)
>> + {
>> + TIME_WAIT = time_wait;
>> + }
>> +
>> + public void UpdateState(TcpPacket packet, bool packetFromInside)
> 
>> What happens if we receive a repeated Syn travelling in the same
>> direction? (From inside to out, say) Is the machine reset? That is,
>> can evil person (call them E) break user A's connection to outside
>> server B, when E sends the NAT a Syn addressed to B (and spoofed to
>> appear to come from A)?
> 
> I'm not sure if I've changed the code since this comment was made. A
> malicious Syn would be ignored, but a malicious Fin would cause the
> NAT entry to be removed after the TIME_WAIT timeout was finished
> regardless of continuing traffic. It would also be easier than normal
> spoofing because the attacker doesn't need to know the sequence number
> for the packet to be accepted by the NAT as valid.
> 
> --
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub [2], or mute the
> thread [3].
> 
> 	*
> 
> Links:
> ------
> [1] https://github.com/niksu/pax/pull/12#discussion_r71153426
> [2]
> https://github.com/niksu/pax/pull/12/files/8f1fd2f3c9e5b30ce2ef0bb72ff888a4a096c87c#r71153426
> [3]
> https://github.com/notifications/unsubscribe-auth/ABDTTgrF3v9d9yjEsq9dgcJUOikTzWOjks5qW4l2gaJpZM4JNj7N

-- 
http://www.cl.cam.ac.uk/~ns441/